flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServerArchiveFetcher.java

```diff
@@ -0,0 +1,244 @@
+
+	private static final JsonFactory jacksonFactory = new JsonFactory();
+	private static final ObjectMapper mapper = new ObjectMapper();
+
+	private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(
+		new ExecutorThreadFactory("Flink-HistoryServer-ArchiveFetcher"));
+	private final JobArchiveFetcherTask fetcherTask;
+	private final long refreshIntervalMillis;
+
+	HistoryServerArchiveFetcher(long refreshIntervalMillis, List<HistoryServer.RefreshLocation> refreshDirs, File webDir) {
+		this.refreshIntervalMillis = refreshIntervalMillis;
+		this.fetcherTask = new JobArchiveFetcherTask(refreshDirs, webDir);
+		if (LOG.isInfoEnabled()) {
+			for (HistoryServer.RefreshLocation refreshDir : refreshDirs) {
+				LOG.info("Monitoring directory {} for archived jobs.", refreshDir.getPath());
+			}
+		}
+	}
+
+	void start() {
+		executor.scheduleWithFixedDelay(fetcherTask, 0, refreshIntervalMillis, TimeUnit.MILLISECONDS);
+	}
+
+	void stop() {
+		executor.shutdown();
+
+		try {
+			if (!executor.awaitTermination(1, TimeUnit.SECONDS)) {
+				executor.shutdownNow();
+			}
+		} catch (InterruptedException ignored) {
+			executor.shutdownNow();
+		}
+	}
+
+	/**
+	 * {@link TimerTask} that polls the directories configured as {@link HistoryServerOptions#HISTORY_SERVER_DIRS} for
+	 * new job archives.
+	 */
+	static class JobArchiveFetcherTask extends TimerTask {
+		private final List<HistoryServer.RefreshLocation> refreshDirs;
+		/** Map containing the JobID of all fetched jobs and the refreshDir from with they originate. */
+		private final Map<String, Path> cachedArchives;
+		private final File webDir;
+		private final File webTmpDir;
+		private final File webJobDir;
+		private final File webOverviewDir;
+
+		private static final String JSON_FILE_ENDING = ".json";
+
+		JobArchiveFetcherTask(List<HistoryServer.RefreshLocation> refreshDirs, File webDir) {
+			this.refreshDirs = refreshDirs;
+			this.cachedArchives = new HashMap<>();
+			this.webDir = webDir;
+			this.webTmpDir = new File(webDir, "tmp");
+			webTmpDir.mkdir();
+			this.webJobDir = new File(webDir, "jobs");
+			webJobDir.mkdir();
+			this.webOverviewDir = new File(webDir, "overviews");
+			webOverviewDir.mkdir();
+		}
+
+		@Override
+		public void run() {
+			try {
+				for (HistoryServer.RefreshLocation refreshLocation : refreshDirs) {
+					Path refreshDir = refreshLocation.getPath();
+					FileSystem refreshFS = refreshLocation.getFs();
+
+					// contents of /:refreshDir
+					FileStatus[] jobArchives;
+					try {
+						jobArchives = refreshFS.listStatus(refreshDir);
+					} catch (IOException e) {
+						LOG.error("Failed to access job archive location for path {}.", refreshDir, e);
+						continue;
+					}
+					if (jobArchives == null) {
+						continue;
+					}
+					boolean updateOverview = false;
+					for (FileStatus jobArchive : jobArchives) {
+						Path jobArchivePath = jobArchive.getPath();
+						String jobID = jobArchivePath.getName();
```

Comment:
I will add the JobID parsing. I would like to keep using a String so that we don't use switch between JobID and String all the time.